indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskLockboxTest.java

```diff
@@ -1272,6 +1276,627 @@ public void testGetLockedIntervalsForEqualPriorityTask() throws Exception
+        Intervals.of("2017/2018"),
+        MEDIUM_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(theLock);
+    Assert.assertFalse(theLock.isRevoked());
+
+    final Set<TaskLock> expectedLocks = ImmutableSet.of(
+        theLock,
+        sharedLock.revokedCopy(),
+        exclusiveLock.revokedCopy(),
+        appendLock.revokedCopy(),
+        replaceLock.revokedCopy()
+    );
+    Assert.assertEquals(expectedLocks, getAllLocks(tasks));
+
+    final Set<TaskLock> expectedActiveLocks = ImmutableSet.of(
+        theLock
+    );
+    Assert.assertEquals(expectedActiveLocks, getAllActiveLocks(tasks));
+  }
+
+  @Test
+  public void testSharedLockCompatibility() throws Exception
+  {
+    final List<Task> tasks = new ArrayList<>();
+
+    final TaskLock theLock = insertTaskLock(
+        TaskLockType.SHARED,
+        Intervals.of("2017/2018"),
+        MEDIUM_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(theLock);
+    Assert.assertFalse(theLock.isRevoked());
+
+    // No overlapping exclusive lock is compatible
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.EXCLUSIVE,
+            Intervals.of("2017-05-01/2017-06-01"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    // Enclosing shared lock of lower priority - compatible
+    final TaskLock sharedLock0 = insertTaskLock(
+            TaskLockType.SHARED,
+            Intervals.of("2016/2019"),
+            LOW_PRIORITY,
+            tasks
+    );
+    Assert.assertNotNull(sharedLock0);
+    Assert.assertFalse(sharedLock0.isRevoked());
+
+    // Enclosed shared lock of higher priority - compatible
+    final TaskLock sharedLock1 = insertTaskLock(
+        TaskLockType.SHARED,
+        Intervals.of("2017-06-01/2017-07-01"),
+        LOW_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(sharedLock1);
+    Assert.assertFalse(sharedLock1.isRevoked());
+
+    // Partially Overlapping shared lock of equal priority - compatible
+    final TaskLock sharedLock2 = insertTaskLock(
+        TaskLockType.SHARED,
+        Intervals.of("2017-05-01/2018-05-01"),
+        MEDIUM_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(sharedLock2);
+    Assert.assertFalse(sharedLock2.isRevoked());
+
+    // Conficting replace locks are incompatible
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.REPLACE,
+            Intervals.of("2017/2018"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    // Conflicting append locks are incompatible
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.APPEND,
+            Intervals.of("2017-05-01/2018-05-01"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    final Set<TaskLock> expectedLocks = ImmutableSet.of(
+        theLock,
+        sharedLock0,
+        sharedLock1,
+        sharedLock2
+    );
+    Assert.assertEquals(expectedLocks, getAllLocks(tasks));
+  }
```

Comment:
Looking over these tests, they follow a very similar pattern.  Along with my commentary about wanting it to be simpler to decipher what the tests are validating, I think you can simplify things quite a bit.  You can make a `TaskLockboxValidator` which is a class that each test would instantiate.  That class can maintain the list of tasks (that you call task).  You can add methods on it like

```
TaskLock expectLockCreated(Type, Interval, priority);

void expectLockNotGranted(Type, Interval, priority);

void expectLockIsNowRevoked(TaskLock);

void validateExpectedLocks(List<TaskLock>);
```

And then if you adjust these tests to be written in terms of that, I think that it will be both easier to understand what the tests are doing *and* simpler to write the tests.

Reply:
Done. Thanks